### PR TITLE
PR: Fix kernel shutdown (IPython console)

### PR DESCRIPTION
**Pull request by impact27**
## Description of Changes

* [ ] Wrote at least one-line docstrings (for any new functions)
* [ ] Added unit test(s) covering the changes (if testable)

* [ ] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

### Issue(s) Resolved

Fixes #

Depends on https://github.com/spyder-ide/spyder/pull/16764
### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct:

### CODE REVIEW

File: spyder/plugins/ipythonconsole/widgets/client.py [hunk 1]
```diff
@@ -108,7 +108,8 @@ def __init__(self, parent, id_,
                  ask_before_closing=False,
                  css_path=None,
                  configuration=None,
-                 handlers={}):
+                 handlers={},
+                 std_dir=None):
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/client.py:112]:
```suggestion
                 handlers={}):
```

I think this change is not necessary for this PR.

File: spyder/plugins/ipythonconsole/widgets/client.py [hunk 2]
```diff
@@ -131,7 +132,7 @@ def __init__(self, parent, id_,
         self.options_button = options_button
         self.history = []
         self.allow_rename = True
-        self.stderr_dir = None
+        self.std_dir = std_dir
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/client.py]:
```suggestion
        self.stderr_dir = None
```

Please revert this change because it's not necessary.

File: spyder/plugins/ipythonconsole/widgets/debugging.py [hunk 3]
```diff
@@ -584,9 +575,6 @@ def _pdb_readline_callback(self, line):
 
     def pdb_input(self, prompt, password=None):
         """Get input for a command."""
-        if self._hidden:
-            raise RuntimeError(
-                'Request for pdb input during hidden execution.')
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/debugging.py:589]: It this related to the changes in jupyter/qtconsole#505?

File: spyder/plugins/ipythonconsole/widgets/main_widget.py [hunk 4]
```diff
@@ -1016,6 +1016,10 @@ def _create_client_for_kernel(self, connection_file, hostname, sshkey,
             cf_path = cf_path if cf_path else None
             connection_file = find_connection_file(filename=cf_filename,
                                                    path=cf_path)
+            if os.path.splitext(connection_file)[1] != ".json":
+                # There might be a file with the same id in the path.
+                connection_file = find_connection_file(
+                    filename=cf_filename + ".json", path=cf_path)
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/main_widget.py:1022]: Why is this necessary for this PR? Or better yet, could you explain why you're adding these lines?

File: spyder/plugins/ipythonconsole/widgets/main_widget.py [hunk 5]
```diff
@@ -1072,17 +1076,13 @@ def _create_client_for_kernel(self, connection_file, hostname, sshkey,
...
                               ask_before_restart=ask_before_restart,
                               css_path=self.css_path,
                               configuration=self.CONFIGURATION,
-                              handlers=self.registered_spyder_kernel_handlers)
-
-        # Change stderr_dir if requested
-        if self._test_dir:
-            client.stderr_dir = self._test_dir
+                              handlers=self.registered_spyder_kernel_handlers,
+                              std_dir=self._test_dir)
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/main_widget.py:1085]:
```suggestion
                              handlers=self.registered_spyder_kernel_handlers)
```

This doesn't seem necessary for this PR.

File: spyder/plugins/ipythonconsole/widgets/main_widget.py [hunk 6]
```diff
@@ -1072,17 +1076,13 @@ def _create_client_for_kernel(self, connection_file, hostname, sshkey,
...
                               show_elapsed_time=show_elapsed_time,
                               reset_warning=reset_warning,
                               ask_before_restart=ask_before_restart,
                               css_path=self.css_path,
                               configuration=self.CONFIGURATION,
-                              handlers=self.registered_spyder_kernel_handlers)
-
-        # Change stderr_dir if requested
-        if self._test_dir:
-            client.stderr_dir = self._test_dir
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/main_widget.py:1085]: Please revert these changes because they are not necessary for this PR, I'd say.

File: spyder/plugins/ipythonconsole/widgets/main_widget.py [hunk 7]
```diff
@@ -1487,11 +1485,8 @@ def create_new_client(self, give_focus=True, filename='', is_cython=False,
                               ask_before_closing=ask_before_closing,
                               css_path=self.css_path,
                               configuration=self.CONFIGURATION,
-                              handlers=self.registered_spyder_kernel_handlers)
-
-        # Change stderr_dir if requested
-        if self._test_dir:
-            client.stderr_dir = self._test_dir
+                              handlers=self.registered_spyder_kernel_handlers,
+                              std_dir=self._test_dir)
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/main_widget.py:1489]:
```suggestion
                              handlers=self.registered_spyder_kernel_handlers)
```

I think this is not necessary here.

File: spyder/plugins/ipythonconsole/widgets/main_widget.py [hunk 8]
```diff
@@ -1487,11 +1485,8 @@ def create_new_client(self, give_focus=True, filename='', is_cython=False,
                               ask_before_closing=ask_before_closing,
                               css_path=self.css_path,
                               configuration=self.CONFIGURATION,
-                              handlers=self.registered_spyder_kernel_handlers)
-
-        # Change stderr_dir if requested
-        if self._test_dir:
-            client.stderr_dir = self._test_dir
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/main_widget.py:1494]: Please revert these changes for now if possible.

File: spyder/plugins/ipythonconsole/widgets/main_widget.py [hunk 9]
```diff
@@ -1777,11 +1767,15 @@ def close_clients(self):
...
-            client.remove_stderr_file()
-            client.dialog_manager.close_all()
+        open_clients = self.clients.copy()
+        for client in self.clients:
+            last_client = (
+                len(self.get_related_clients(client, open_clients)) == 0)
+            client.shutdown(last_client)
             client.close()
+            open_clients.remove(client)
+        # close all closing shellwidgets
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/main_widget.py:1777]:
```suggestion
        # Close all closing shellwidgets.
```

File: spyder/plugins/ipythonconsole/widgets/namespacebrowser.py [hunk 10]
```diff
@@ -205,7 +205,7 @@ def _handle_execute_reply(self, msg):
             self.ipyclient.t0 = time.monotonic()
 
         # Handle silent execution of kernel methods
-        if info and info.kind == 'silent_exec_method' and not self._hidden:
+        if info and info.kind == 'silent_exec_method':
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/namespacebrowser.py:208]: Does this need jupyter/qtconsole#505?

File: spyder/plugins/ipythonconsole/widgets/shell.py [hunk 11]
```diff
@@ -77,6 +78,27 @@ class ShellWidget(NamepaceBrowserWidget, HelpWidget, DebuggingWidget,
     # For printing internal errors
     sig_exception_occurred = Signal(dict)
 
+    # Class array of shutdown threads
+    shutdown_thread_list = []
+
+    @classmethod
+    def prune_shutdown_thread_list(cls):
+        """Remove  shutdown threads."""
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/shell.py:86]:
```suggestion
        """Remove shutdown threads."""
```

File: spyder/plugins/outlineexplorer/tests/test_widgets.py [hunk 12]
```diff
@@ -267,7 +267,7 @@ def test_go_to_cursor_position_with_new_file(create_outlineexplorer, qtbot):
     # Click on the 'Go to cursor position' button of the outline explorer's
     # toolbar :
     filename = CASES['text']['file']
-    assert outlineexplorer.treewidget.currentItem() is None
+    # assert outlineexplorer.treewidget.currentItem() is None
```
Comment (ccordoba12) [inline spyder/plugins/outlineexplorer/tests/test_widgets.py:270]: If you consider this is unnecessary, please remove it (I saw your comment on PR #14025).

File: spyder/plugins/outlineexplorer/tests/test_widgets.py [hunk 13]
```diff
@@ -293,7 +293,7 @@ def test_go_to_last_item(create_outlineexplorer, qtbot):
 
     # Click on the 'Go to cursor position' button of the outline explorer's
     # toolbar :
-    assert outlineexplorer.treewidget.currentItem() is None
+    # assert outlineexplorer.treewidget.currentItem() is None
```
Comment (ccordoba12) [inline spyder/plugins/outlineexplorer/tests/test_widgets.py:296]: Also remove this line.

**Review by ccordoba12 — COMMENTED** [review]
@impact27, thanks a lot for your help with this one!

**Comment by ccordoba12** [comment]
The failures in our Mac app tests are unrelated to this. The base of our PRs need to be the right one for they to pass (since I moved it to `5.x`, the next commit you push should fix them).

File: spyder/plugins/ipythonconsole/widgets/debugging.py (hunk 3 above)
Comment (impact27) [reply spyder/plugins/ipythonconsole/widgets/debugging.py:589]: Yes because self._hidden is only meaningful with relation to a specific message which we don't have access to here.

File: spyder/plugins/ipythonconsole/widgets/main_widget.py (hunk 4 above)
Comment (impact27) [reply spyder/plugins/ipythonconsole/widgets/main_widget.py:1022]: So when testing, the stderr file (and in the future, the .stdout and .fault) are returned by `find_connection_file`. This is because this code was moved from the plugin folder to the widget folder I think.

Comment (impact27) [reply spyder/plugins/ipythonconsole/widgets/main_widget.py:1022]: The idea is that find_connection_file searches for a file with name matching the kernel ID. unfortunately the stderr file matches this so can be returned instead of the json file.

File: spyder/plugins/ipythonconsole/widgets/debugging.py (hunk 3 above)
Comment (ccordoba12) [reply spyder/plugins/ipythonconsole/widgets/debugging.py:589]: Ok, I'll merge that PR and release 5.1.2 with the fix so that our tests pass.

File: spyder/plugins/ipythonconsole/widgets/namespacebrowser.py (hunk 10 above)
Comment (impact27) [reply spyder/plugins/ipythonconsole/widgets/namespacebrowser.py:208]: jupyter/qtconsole#505 needs this but not the other way around. Anyway you should never have _hidden == True when info.kind == 'silent_exec_method'

File: spyder/plugins/outlineexplorer/tests/test_widgets.py (hunk 12 above)
Comment (impact27) [reply spyder/plugins/outlineexplorer/tests/test_widgets.py:270]: I am 95% sure this is unnecessary. I need to remove it to pass the test but I don't think the failure is related to this PR.

**Comment by ccordoba12** [comment]
Thanks a lot for your help @impact27! I couldn't release qtconsole 5.2.0 yesterday, just some hours ago.

But now I'm fixing our dependency on it in conda-forge (see https://github.com/conda-forge/spyder-feedstock/pull/119) because 5.1.5 depends on `self._hidden` being present in `RichJupyterWidget` and the new qtconsole removed it.

**Comment by impact27** [comment]
_hidden is remove here: https://github.com/spyder-ide/spyder/pull/16764

**Comment by ccordoba12** [comment]
@impact27, since #16764 is now in, could you rebase this one so that it doesn't contain the commits present in that PR? Thanks!

**Review by ccordoba12 — APPROVED** [review]
Looks good to me now. Thanks a lot for your hard work on this one @impact27!
